IMPORTING THE DATA

In [2]:
import pandas as pd

In [3]:
matches = pd.read_csv("Documents/matches.csv",index_col=0)

In [4]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [5]:
matches.shape

(1389, 27)

INVESTIGATING MISSING DATA

***So if you know the EPL, there are 38 matches played by each team, and there a total of 20 teams in the EPL. We have scraaped the data of 2 seasons, namely the 20/21 and the 21/22 seasons.
This means we should have 1520 (38 * 20 * 2) teams.
But as we saw above, we only have 1389 rows meaning there is a lot of missing data we need to account for.***

***Let us first check each team that have played in these 2 seasons and list out the total number of matches that they have played.***

In [10]:
matches["team"].value_counts()

Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: team, dtype: int64

***Now, if you follow the EPL closely, you know that 3 teams get relegated to the Championship each season. This implies that we will have 6 teams that must have played 38 matches and all the other teams must have played 72 (38 * 2) matches. But if you observe closely, we have 7 teams that have played around 38 matches, this indicates that one team has over half of their matches missing***

***If you observe closely, that team is Liverpool. It is a strong team that has not been relegated for more than the past 3 decades. And we only have 38 matches of Liverpool meaning we are essentially missing an entire season of games for LFC***

In [11]:
matches[matches["team"] == "Liverpool"]

date   time            comp         round  day venue result   gf  \
1   2020-09-12  17:30  Premier League   Matchweek 1  Sat  Home      W  4.0   
2   2020-09-20  16:30  Premier League   Matchweek 2  Sun  Away      W  2.0   
4   2020-09-28  20:00  Premier League   Matchweek 3  Mon  Home      W  3.0   
6   2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7   2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9   2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11  2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13  2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14  2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16  2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18  2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20  2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21  2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22  2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23  2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24  2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25  2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27  2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28  2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30  2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31  2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32  2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33  2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34  2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36  2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37  2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38  2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39  2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41  2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42  2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44  2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46  2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47  2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48  2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49  2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50  2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51  2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  match report  notes    sh   sot  dist   fk  \
1   3.0     Leeds United  ...  Match Report    NaN  20.0   4.0  17.0  0.0   
2   0.0          Chelsea  ...  Match Report    NaN  17.0   5.0  17.7  1.0   
4   1.0          Arsenal  ...  Match Report    NaN  21.0   9.0  16.8  0.0   
6   7.0      Aston Villa  ...  Match Report    NaN  14.0   8.0  15.8  1.0   
7   2.0          Everton  ...  Match Report    NaN  22.0   8.0  15.0  1.0   
9   1.0    Sheffield Utd  ...  Match Report    NaN  17.0   5.0  18.2  1.0   
11  1.0         West Ham  ...  Match Report    NaN   8.0   2.0  18.6  1.0   
13  1.0  Manchester City  ...  Match Report    NaN   9.0   2.0  21.5  0.0   
14  0.0   Leicester City  ...  Match Report    NaN  24.0  12.0  11.9  0.0   
16  1.0         Brighton  ...  Match Report    NaN   6.0   2.0  20.9  0.0   
18  0.0           Wolves  ...  Match Report    NaN  11.0   6.0  16.6  1.0   
20  1.0           Fulham  ...  Match Report

In [12]:
matches["round"].value_counts()

Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: round, dtype: int64

***Some of the Matchweeks have fewer than 39 matches because of two reasons: 1.)We have many games of Liverpool missing 2.)We scraped the data when the 21/22 season was still going on, so we may have some data missing from there as well***

CLEANING THE DATA

In [14]:
matches.dtypes      #Trying to understand the data types of our features

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [16]:
matches["date"] = pd.to_datetime(matches["date"])     #overwriting date column from string to date-time dtype

In [22]:
matches.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

CREATING BASIC PREDICTORS AS A BASE

***We'll create a Venue Code which basically converts the string "Home" or "Awsy" into a numerical code so that our algorithm can work on them and include the factor of "Home Ground Advantage"***

In [24]:
matches["venue_code"] = matches["venue"].astype("category").cat.codes         #Away:0 and Home:1

In [26]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  notes    sh   sot  dist   fk   pk pkatt season  \
1       Tottenham  ...    NaN  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2    Norwich City  ...    NaN  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3         Arsenal  ...    NaN  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4  Leicester City  ...    NaN  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6     Southampton  ...    NaN  16.0   1.0  15.7  1.0  0.0   0.0   2022   

              team  venue_code  
1  Manchester City           0  
2  Manchester City           1  
3  Manchester City           1  
4  Manchester City           0  
6  Manchester City           1  

[5 rows x 28 columns]

***Similarily, we will also convert the opposition team names into numerical values as well so that we can factor them into our algorithm***

In [28]:
matches["opp_code"] = matches["opponent"].astype("category").cat.codes

In [29]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...    sh   sot  dist   fk   pk pkatt season  \
1       Tottenham  ...  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2    Norwich City  ...  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3         Arsenal  ...  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4  Leicester City  ...  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6     Southampton  ...  16.0   1.0  15.7  1.0  0.0   0.0   2022   

              team  venue_code  opp_code  
1  Manchester City           0        18  
2  Manchester City           1        15  
3  Manchester City           1         0  
4  Manchester City           0        10  
6  Manchester City           1        17  

[5 rows x 29 columns]

***Some teams may play better at a particular time and we need to include that as well. We will be writing a Regex to extract the hour from the time column as the minutes part does not make a huge difference to our model. For eg: If the kick-off time is "16:30" we well extract the "16" part and discard the ":30" part.***

In [31]:
matches["hour"] = matches["time"].str.replace(":.+", "", regex=True).astype("int")

In [32]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...   sot  dist   fk   pk pkatt season             team  \
1       Tottenham  ...   4.0  16.9  1.0  0.0   0.0   2022  Manchester City   
2    Norwich City  ...   4.0  17.3  1.0  0.0   0.0   2022  Manchester City   
3         Arsenal  ...  10.0  14.3  0.0  0.0   0.0   2022  Manchester City   
4  Leicester City  ...   8.0  14.0  0.0  0.0   0.0   2022  Manchester City   
6     Southampton  ...   1.0  15.7  1.0  0.0   0.0   2022  Manchester City   

  venue_code  opp_code  hour  
1          0        18    16  
2          1        15    15  
3          1         0    12  
4          0        10    15  
6          1        17    15  

[5 rows x 30 columns]

***We will encode the days into numbers as well. If you follow the EPL closely, you'll know that during the winters, there is an overload of matches and hence each team has to play twice a week, once on Wednesday/Thursday and once on Saturday/Sunday. This may affect the performance of certain teams and that is why it becomes an important feature for us to consider.***

In [33]:
matches["day_code"] = matches["date"].dt.dayofweek

In [34]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  dist   fk   pk  pkatt season             team  \
1       Tottenham  ...  16.9  1.0  0.0    0.0   2022  Manchester City   
2    Norwich City  ...  17.3  1.0  0.0    0.0   2022  Manchester City   
3         Arsenal  ...  14.3  0.0  0.0    0.0   2022  Manchester City   
4  Leicester City  ...  14.0  0.0  0.0    0.0   2022  Manchester City   
6     Southampton  ...  15.7  1.0  0.0    0.0   2022  Manchester City   

  venue_code opp_code  hour  day_code  
1          0       18    16         6  
2          1       15    15         5  
3          1        0    12         5  
4          0       10    15         5  
6          1       17    15         5  

[5 rows x 31 columns]

In [37]:
matches["target"] = (matches["result"] == "W").astype("int")    #Establishing our target variable where '0' is a loss and '1' is a win

In [38]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...   fk   pk  pkatt  season             team venue_code  \
1       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City          0   
2    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City          1   
3         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City          1   
4  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City          0   
6     Southampton  ...  1.0  0.0    0.0    2022  Manchester City          1   

  opp_code hour  day_code  target  
1       18   16         6       0  
2       15   15         5       1  
3        0   12         5       1  
4       10   15         5       1  
6       17   15         5       0  

[5 rows x 32 columns]

CREATING BASIC MACHINE LEARNING MODEL

In [39]:
from sklearn.ensemble import RandomForestClassifier

***We will be using a Random Forest Classifier as it is a Machine Learning Model famously known for its ability to pickup the non-lineraity in the data. As you can observe in the data, the "opp_code" has numbers from 0-19, but that does not imply that the opposition with the code 18 is more difficult than the opposition with the code 5. Now this can be picked up by the Non-Linear ML models but cannot be identified by the Linear ML models.***

In [225]:
rf = RandomForestClassifier(n_estimators=200, min_samples_split=10, random_state=1)

***The "n_estimators" parameter basically defines the number of individual decision trees but each tree has slightly different parameters. Basically, higher the number more the time it will take to train the model but will also ensure a higher accuracy. The "min_samples_split" parameter defines the minimum number of ssamples we want to have in a leaf node of a decision tree before we actually split it. The bhigher this is we will be less likely to overfit but also have a lower accuracy on the training dataset, so we'll need to experiment with it's value.***

In [226]:
train = matches[matches["date"] < "2022-01-01"]    #subsetting our training data from the original data

In [227]:
test = matches[matches["date"] > "2022-01-01"]   #subsetting our test data from the original data

In [228]:
predictors = ["venue_code", "opp_code", "hour", "day_code"]  #list of all our defined predictors

In [229]:
rf.fit(train[predictors], train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=200, random_state=1)

***We have fit our model with the training dataset. In the rf.fit(x, y), the x part suggest on which data we want the model to train on and the y part suggests what is the target that we need to predicr with the help our training data.***

In [230]:
preds = rf.predict(test[predictors])

***To determine the accuracy of the matrix, we will be importing a couple of different matrics and then figure out which one works best for our case.***

In [231]:
from sklearn.metrics import accuracy_score

In [232]:
acc = accuracy_score(test["target"], preds)

In [233]:
acc

0.6123188405797102

***Now let us understand where the actual and predicted valuse were different so that we could further optimize our model.***

In [54]:
combined = pd.DataFrame(dict(actual=test["target"], prediction=preds))

In [57]:
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction    0   1
actual             
0           142  30
1            78  26

In [58]:
from sklearn.metrics import precision_score

In [59]:
precision_score(test["target"], preds)

0.4642857142857143

***This prediction score tells us that out of the total times we predicted a Win, what percentage of time we were right.***

IMPROVING PRECISION WITH ROLLING AVERAGES

In [60]:
grouped_matches = matches.groupby("team")

In [61]:
group = grouped_matches.get_group("Manchester City")

In [62]:
group

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
54 2021-05-01  12:30  Premier League  Matchweek 34  Sat  Away      W  2.0   
56 2021-05-08  17:30  Premier League  Matchweek 35  Sat  Home      L  1.0   
57 2021-05-14  20:00  Premier League  Matchweek 36  Fri  Away      W  4.0   
58 2021-05-18  19:00  Premier League  Matchweek 37  Tue  Away      L  2.0   
59 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  5.0   

     ga        opponent  ...   fk   pk  pkatt  season             team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022  Manchester City   
..  ...             ...  ...  ...  ...    ...     ...              ...   
54  0.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Manchester City   
56  2.0         Chelsea  ...  0.0  0.0    1.0    2021  Manchester City   
57  3.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Manchester City   
58  3.0        Brighton  ...  1.0  0.0    0.0    2021  Manchester City   
59  0.0         Everton  ...  0.0  0.0    0.0    2021  Manchester City   

   venue_code opp_code hour  day_code  target  
1           0       18   16         6       0  
2           1       15   15         5       1  
3           1        0   12         5       1  
4           0       10   15         5       1  
6           1       17   15         5       0  
..        ...      ...  ...       ...     ...  
54          0        6   12         5       1  
56          1        5   17         5       0  
57          0       14   20         4       1  
58          0        3   19         1       0  
59          1        7   16         6       1  

[71 rows x 32 columns]

***So, basically what we did here is that we grouped all the matches played by a single team into one group so that when we make prediction we can look at their past matches, shots on target, goals scored goals conceded and many more factors. This helps us include the sports term "run-of-form" into our algorithm. This is a very important feature as a team might not be that good on paper, but it may have a good run of form which heavily influences the performance of that team.***

In [174]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    rolling_stats = group[cols].rolling(3, closed="left").mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group

In [175]:
cols = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]

In [176]:
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [177]:
rolling_averages(group, cols, new_cols)  #This is for Man City only as we had called the group function for ManCity above

date   time            comp         round  day venue result   gf  \
5  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Home      W  1.0   
7  2020-10-24  12:30  Premier League   Matchweek 6  Sat  Away      D  1.0   
9  2020-10-31  12:30  Premier League   Matchweek 7  Sat  Away      W  1.0   
11 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Home      D  1.0   
12 2020-11-21  17:30  Premier League   Matchweek 9  Sat  Away      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
42 2022-03-14  20:00  Premier League  Matchweek 29  Mon  Away      D  0.0   
44 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Away      W  2.0   
46 2022-04-10  16:30  Premier League  Matchweek 32  Sun  Home      D  2.0   
49 2022-04-20  20:00  Premier League  Matchweek 30  Wed  Home      W  3.0   
50 2022-04-23  15:00  Premier League  Matchweek 34  Sat  Home      W  5.0   

     ga        opponent  ...  day_code  target  gf_rolling  ga_rolling  \
5   0.0         Arsenal  ...         5       1    2.000000    2.333333   
7   1.0        West Ham  ...         5       0    1.333333    2.000000   
9   0.0   Sheffield Utd  ...         5       1    1.000000    0.666667   
11  1.0       Liverpool  ...         6       0    1.000000    0.333333   
12  2.0       Tottenham  ...         5       0    1.000000    0.666667   
..  ...             ...  ...       ...     ...         ...         ...   
42  0.0  Crystal Palace  ...         0       0    2.333333    1.333333   
44  0.0         Burnley  ...         5       1    1.666667    0.333333   
46  2.0       Liverpool  ...         6       0    2.000000    0.333333   
49  0.0        Brighton  ...         2       1    1.333333    0.666667   
50  1.0         Watford  ...         5       1    2.333333    0.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5   17.333333    4.666667    18.900000   1.333333    0.333333       0.333333  
7   17.333333    3.666667    17.733333   0.666667    0.000000       0.000000  
9   16.666667    4.333333    18.233333   0.666667    0.000000       0.000000  
11  14.333333    6.666667    18.466667   1.000000    0.000000       0.000000  
12  12.000000    5.666667    19.366667   1.000000    0.000000       0.333333  
..        ...         ...          ...        ...         ...            ...  
42  19.000000    7.000000    15.366667   0.333333    0.333333       0.333333  
44  18.333333    7.333333    16.000000   0.333333    0.000000       0.000000  
46  20.000000    6.666667    16.133333   0.333333    0.000000       0.000000  
49  15.666667    4.666667    16.700000   0.333333    0.000000       0.000000  
50  15.333333    5.000000    17.200000   0.333333    0.000000       0.000000  

[68 rows x 40 columns]

***As we can observe from the Manchester City group, we have added the rolling averages and have a lot of additional information about the team's  performance and will be able to make better predictions. So, we will now group all the other teams and appy the same rolling_averages function to their groups using lambda functions.***

In [178]:
matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))

In [179]:
matches_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 6          6       1    2.000000    1.333333   
                        7          5       0    1.666667    1.666667   
                        9          6       0    1.000000    1.666667   
                        11         6       1    0.666667    1.000000   
                        13         6       0    0.333333    0.666667   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 32         6       1    1.333333    1.000000   
                        33         4       0    1.666667    0.666667   
                        34         5       1    2.333333    1.000000   
                        35         4       0    1.666667    1.333333   
                        36         6       0    1.333333    1.666667   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 6    7.666667    3.666667    14.733333   0.666667   
                        7    5.333333    3.666667    15.766667   0.000000   
                        9    7.000000    3.666667    16.733333   0.666667   
                        11   9.666667    4.000000    16.033333   1.000000   
                        13   9.666667    2.666667    18.033333   1.000000   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 32  12.333333    3.666667    19.300000   0.000000   
                        33  12.333333    4.333333    19.600000   0.000000   
                        34  13.000000    5.333333    19.833333   0.000000   
                        35  13.000000    5.000000    18.533333   0.000000   
                        36  10.000000    4.666667    17.633333   0.000000   

                            pk_rolling

In [180]:
matches_rolling = matches_rolling.droplevel("team")

In [181]:
matches_rolling

date   time            comp         round  day venue result   gf  \
6  2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
7  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
9  2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
11 2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
13 2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
32 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
33 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
34 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
35 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
36 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
6   1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
7   1.0  Manchester City  ...         5       0    1.666667    1.666667   
9   1.0   Leicester City  ...         6       0    1.000000    1.666667   
11  0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
13  3.0      Aston Villa  ...         6       0    0.333333    0.666667   
..  ...              ...  ...       ...     ...         ...         ...   
32  0.0          Everton  ...         6       1    1.333333    1.000000   
33  3.0     Leeds United  ...         4       0    1.666667    0.666667   
34  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
35  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
36  1.0          Burnley  ...         6       0    1.333333    1.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
6    7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
7    5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
9    7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
11   9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
13   9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
..        ...         ...          ...        ...         ...            ...  
32  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
33  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
34  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
35  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
36  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

In [182]:
matches_rolling.index = range(matches_rolling.shape[0])

***So what we basically did here is that we assigned a unique id to all the rows so tha we don not have to deal with redundant row indexes which makes it easier for us to call and handle the rows.***

In [183]:
matches_rolling

date   time            comp         round  day venue result   gf  \
0    2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
1    2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
2    2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
3    2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
4    2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
1312 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
1313 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
1314 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
1315 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
1316 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

       ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
0     1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
1     1.0  Manchester City  ...         5       0    1.666667    1.666667   
2     1.0   Leicester City  ...         6       0    1.000000    1.666667   
3     0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
4     3.0      Aston Villa  ...         6       0    0.333333    0.666667   
...   ...              ...  ...       ...     ...         ...         ...   
1312  0.0          Everton  ...         6       1    1.333333    1.000000   
1313  3.0     Leeds United  ...         4       0    1.666667    0.666667   
1314  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
1315  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
1316  1.0          Burnley  ...         6       0    1.333333    1.666667   

     sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0      7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
1      5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
2      7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
3      9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
4      9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
...         ...         ...          ...        ...         ...            ...  
1312  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
1313  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
1314  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
1315  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
1316  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

RETRAINING OUR MODEL

In [184]:
def make_predictions(data, predictors):
    train = data[data["date"] < "2022-01-01"]
    test = data[data["date"] > "2022-01-01"]
    rf.fit(train[predictors], train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)
    precision = precision_score(test["target"], preds)
    return combined, precision

In [185]:
combined, precision = make_predictions(matches_rolling, predictors + new_cols)

In [186]:
precision

0.625

***Increased the precision from 46.4% to 62.5% !!***

In [187]:
combined

actual  predicted
55         0          0
56         1          0
57         1          0
58         1          1
59         1          1
...      ...        ...
1312       1          0
1313       0          0
1314       1          0
1315       0          0
1316       0          0

[276 rows x 2 columns]

In [188]:
combined = combined.merge(matches_rolling[["date", "team", "opponent", "result"]], left_index=True, right_index=True)

In [189]:
combined

actual  predicted       date                     team       opponent  \
55         0          0 2022-01-23                  Arsenal        Burnley   
56         1          0 2022-02-10                  Arsenal         Wolves   
57         1          0 2022-02-19                  Arsenal      Brentford   
58         1          1 2022-02-24                  Arsenal         Wolves   
59         1          1 2022-03-06                  Arsenal        Watford   
...      ...        ...        ...                      ...            ...   
1312       1          0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0          0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1          0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0          0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0          0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

COMBINING HOME AND AWAY PREDICTIONS

***As we can see above that in a particular group we can only see the predictions for the group team only. We shall also include the predictions for the other teams as well so that it can be easier to read for our algorithm and it will also be easier on the eyes for the programmer, making it easy to read and understand the data.***

In [190]:
class MissingDict(dict):
    __missing__ = lambda self, key: key
    
map_values = {
    "Brighton and Hove ALbion": "Brighton",
    "Manchester United": "Manchester Utd",
    "Newcastle United": "Newcastle Utd",
    "Tottenham Hotspur": "Tottenham",
    "West Ham United": "West Ham",
    "Wolverhampton Wanderers": "Wolves"
}
mapping = MissingDict(**map_values)

In [191]:
combined["new_team"] = combined["team"].map(mapping)

In [192]:
combined

actual  predicted       date                     team       opponent  \
55         0          0 2022-01-23                  Arsenal        Burnley   
56         1          0 2022-02-10                  Arsenal         Wolves   
57         1          0 2022-02-19                  Arsenal      Brentford   
58         1          1 2022-02-24                  Arsenal         Wolves   
59         1          1 2022-03-06                  Arsenal        Watford   
...      ...        ...        ...                      ...            ...   
1312       1          0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0          0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1          0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0          0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0          0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result new_team  
55        D  Arsenal  
56        W  Arsenal  
57        W  Arsenal  
58        W  Arsenal  
59        W  Arsenal  
...     ...      ...  
1312      W   Wolves  
1313      L   Wolves  
1314      W   Wolves  
1315      L   Wolves  
1316      L   Wolves  

[276 rows x 7 columns]

In [193]:
merged = combined.merge(combined, left_on=["date", "new_team"], right_on=["date","opponent"])

In [194]:
merged

actual_x  predicted_x       date                   team_x     opponent_x  \
0           0            0 2022-01-23                  Arsenal        Burnley   
1           1            0 2022-02-10                  Arsenal         Wolves   
2           1            0 2022-02-19                  Arsenal      Brentford   
3           1            1 2022-02-24                  Arsenal         Wolves   
4           1            1 2022-03-06                  Arsenal        Watford   
..        ...          ...        ...                      ...            ...   
242         1            0 2022-03-13  Wolverhampton Wanderers        Everton   
243         0            0 2022-03-18  Wolverhampton Wanderers   Leeds United   
244         1            0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
245         0            0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
246         0            0 2022-04-24  Wolverhampton Wanderers        Burnley   

    result_x new_team_x  actual_y  predicted_y                   team_y  \
0          D    Arsenal         0            0                  Burnley   
1          W    Arsenal         0            0  Wolverhampton Wanderers   
2          W    Arsenal         0            0                Brentford   
3          W    Arsenal         0            0  Wolverhampton Wanderers   
4          W    Arsenal         0            0                  Watford   
..       ...        ...       ...          ...                      ...   
242        W     Wolves         0            0                  Everton   
243        L     Wolves         1            0             Leeds United   
244        W     Wolves         0            0              Aston Villa   
245        L     Wolves         1            0         Newcastle United   
246        L     Wolves         1            0                  Burnley   

    opponent_y result_y     new_team_y  
0      Arsenal        D        Burnley  
1      Arsenal        L         Wolves  
2      Arsenal        L      Brentford  
3      Arsenal        L         Wolves  
4      Arsenal        L        Watford  
..         ...      ...            ...  
242     Wolves        L        Everton  
243     Wolves        W   Leeds United  
244     Wolves        L    Aston Villa  
245     Wolves        W  Newcastle Utd  
246     Wolves        W        Burnley  

[247 rows x 13 columns]

In [195]:
merged[(merged["predicted_x"] == 1) & (merged["predicted_y"] == 0)]["actual_x"].value_counts()

1    27
0    13
Name: actual_x, dtype: int64

In [196]:
precision = 27/(27+13)

In [197]:
precision

0.675